src/clr-angular/data/datagrid/datagrid-row-detail.ts

```diff
@@ -49,14 +64,57 @@ export class ClrDatagridRowDetail<T = any> implements AfterContentInit, OnDestro
   public replacedRow = false;
 
   ngAfterContentInit() {
-    this.subscriptions.push(
-      this.expand.replace.subscribe(replaceChange => {
-        this.replacedRow = replaceChange;
-      })
-    );
+    this.subscriptions.push(this.reorderViewsOnOrdersChange(), this.setReplacedOnExpandReplace());
+  }
+
+  private setReplacedOnExpandReplace(): Subscription {
+    return this.expand.replace.subscribe(replaceChange => {
+      this.replacedRow = replaceChange;
+    });
   }
 
   ngOnDestroy() {
     this.subscriptions.forEach(sub => sub.unsubscribe());
   }
+
+  private reorderViewsOnOrdersChange(): Subscription {
+    return this.columnReorderService.ordersChange.subscribe((byReordering: boolean) => {
+      // we have to make sure there are views inserted in the view container
+      // before calling reorderViews method from viewManager service
+      if (byReordering && this._detailCells.length > 0) {
+        ViewManagerUtils.reorderViews(this._detailCells, this.assignRawOrders());
+        this.updateCellOrder();
+        return;
+      }
+
+      if (this.columnReorderService.orders) {
+        if (this._detailCells.length === 0) {
+          // directly insert the views initially if the view container is empty.
+          ViewManagerUtils.insertAllViews(this._detailCells, this.assignRawOrders(), true);
+        } else {
+          this.cells.forEach((cell, index) => {
+            if (this._detailCells.indexOf(cell._view) === -1) {
+              this._detailCells.insert(cell._view, this.columnReorderService.orderAt(index));
+            }
+          });
+        }
+        this.updateCellOrder();
+      }
+    });
+  }
+
+  private updateCellOrder(): void {
+    this.cells.forEach(cell => (cell.order = this._detailCells.indexOf(cell._view)));
+  }
+
+  private assignRawOrders(): ClrDatagridCell[] {
```

Comment:
Some of this is duplicated, can we share is somehow?

Reply:
Yes, this method is in ClrDatagrid, ClrDatagridRow, ClrDatagridRowDetail. Only ClrDatagridRow and ClrDatagridRowDetail share the exact same implementation. So the implementation is not exactly the same across these classes. So that's also why I didn't turn it into a utility method or something that's shared.